pkg/controller/endpointslicemirroring/reconciler.go

```diff
@@ -0,0 +1,287 @@
+	// the corresponding endpoint slices for deletion.
+	for portKey, existingSlices := range existingSlicesByKey {
+		if _, ok := d.endpointsByKey[portKey]; !ok {
+			for _, existingSlice := range existingSlices {
+				slices.toDelete = append(slices.toDelete, existingSlice)
+			}
+		}
+	}
+
+	metrics.EndpointsAddedPerSync.WithLabelValues().Observe(float64(totals.added))
+	metrics.EndpointsUpdatedPerSync.WithLabelValues().Observe(float64(totals.updated))
+	metrics.EndpointsRemovedPerSync.WithLabelValues().Observe(float64(totals.removed))
+
+	endpointsNN := types.NamespacedName{Name: endpoints.Name, Namespace: endpoints.Namespace}
+	r.metricsCache.UpdateEndpointPortCache(endpointsNN, epMetrics)
+
+	return r.finalize(endpoints, slices)
+}
+
+// reconcileByPortMapping compares the endpoints found in existing slices with
+// the list of desired endpoints and returns lists of slices to create, update,
+// and delete.
+func (r *reconciler) reconcileByPortMapping(
+	endpoints *corev1.Endpoints,
+	existingSlices []*discovery.EndpointSlice,
+	desiredSet endpointSet,
+	endpointPorts []discovery.EndpointPort,
+	addressType discovery.AddressType,
+) (slicesByAction, totalsByAction) {
+	slices := slicesByAction{}
+	totals := totalsByAction{}
+
+	// If no endpoints are desired, mark existing slices for deletion and
+	// return.
+	if desiredSet.Len() == 0 {
+		slices.toDelete = existingSlices
+		for _, epSlice := range existingSlices {
+			totals.removed += len(epSlice.Endpoints)
+		}
+		return slices, totals
+	}
+
+	if len(existingSlices) == 0 {
+		// if no existing slices, all desired endpoints will be added.
+		totals.added = desiredSet.Len()
+	} else {
+		// if >0 existing slices, mark all but 1 for deletion.
+		slices.toDelete = existingSlices[1:]
+
+		// Return early if first slice matches desired endpoints.
+		totals = totalChanges(existingSlices[0], desiredSet)
+		if totals.added == 0 && totals.updated == 0 && totals.removed == 0 {
+			return slices, totals
+		}
+	}
+
+	// generate a new slice with the desired endpoints.
+	newSlice := newEndpointSlice(endpoints, endpointPorts, addressType)
+	for desiredSet.Len() > 0 && len(newSlice.Endpoints) < int(r.maxEndpointsPerSubset) {
+		endpoint, _ := desiredSet.PopAny()
+		newSlice.Endpoints = append(newSlice.Endpoints, *endpoint)
+	}
+
+	if len(existingSlices) > 0 {
+		newSlice.Name = existingSlices[0].Name
+		slices.toUpdate = []*discovery.EndpointSlice{newSlice}
+	} else {
+		slices.toCreate = []*discovery.EndpointSlice{newSlice}
+	}
+	return slices, totals
+}
+
+// finalize creates, updates, and deletes slices as specified
+func (r *reconciler) finalize(endpoints *corev1.Endpoints, slices slicesByAction) error {
+	errs := []error{}
+
+	// If there are slices to create and delete, recycle the slices marked for
+	// deletion by replacing creates with updates of slices that would otherwise
+	// be deleted.
+	recycleSlices(&slices)
+
+	// Don't create more EndpointSlices if corresponding Endpoints resource is
```

Comment:
might make below easier to read

`epsClient := r.client.DiscoveryV1beta1().EndpointSlices(endpoints.Namespace)`

Reply:
Updated, thanks!